File: MapAccounts/MapAccounts/Extensions/Vector2D.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace MapAccounts.Extensions
{
    public class Vector2D
    {
        private double x;
        private double y;

        public double X { get { return x; } }
        public double Y { get { return y; } }

        public Vector2D(double _x, double _y)
        {
            x = _x;
            y = _y;
        }

        public static Vector2D operator +(Vector2D a, Vector2D b)
        {
            return new Vector2D(a.x + b.x, a.y + b.y);
        }
        public static Vector2D operator -(Vector2D a, Vector2D b)
        {
            return new Vector2D(a.x - b.x, a.y - b.y);
        }
        public static Vector2D operator *(Vector2D a, double c)
        {
            return new Vector2D(a.x*c, a.y*c);
        }
        public static double dot(Vector2D a, Vector2D b)
        {
            return (a.x * b.x) + (a.y * b.y);
        }
        public static double norm(Vector2D a)
        {
            return Math.Sqrt((a.x * a.x) + (a.y * a.y));
        }
        public static double angleBetween(Vector2D a, Vector2D b)
        {
            var na = normalized(a);
            var nb = normalized(b);

            var d = Vector2D.dot(na, nb);

            return (d >= 0) ? Math.Acos(d) : Math.PI*2.0 - Math.Acos(d);

        }
        public static Vector2D normalized(Vector2D a)
        {
            double norm = Vector2D.norm(a);
            if (norm == 0) throw new DivideByZeroException("Vector should be non-zero!");
            return a * (1.0 / norm);
        }
    }
}

File: MapAccounts/MapAccounts/Client/Scripts/Helpers/Vector2D.js
function Vector2D()
{
    var x;
    var y;

    this.X = function() { return x;  }
    this.Y = function() { return y;  }
    this.initialize = function(_x, _y)
    {
        x = _x;
        y = _y;
    }
}

Vector2D.add = function(a, b) { var v = new Vector2D(); v.initialize(a.X+b.X, a.Y+b.Y); return v;}
Vector2D.sub = function(a, b) { var v = new Vector2D(); v.initialize(a.X-b.X, a.Y-b.Y); return v;}
Vector2D.add = function(a, c) { var v = new Vector2D(); v.initialize(a.X*c, a.Y*c); return v;}
Vector2D.dot = function(a, b) {  return (a.X*b.X) + (a.Y*b.Y); }
Vector2D.norm = function(a) {  return Math.sqrt((a.x * a.x) + (a.y * a.y)); }